lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1445,6 +1445,9 @@ def _undelete(self, item: HistoryItemModel):
             raise exceptions.ItemDeletionException("This item has been permanently deleted and cannot be recovered.")
         manager = self._get_item_manager(item)
         manager.undelete(item, flush=self.flush)
+        # Again, we need to force an update in the edge case where all selected items are already undeleted
+        # or when the item was purged as undelete will not trigger an update
+        item.update()
```

Comment:
This should probably also go into the undelete method ? If we didn't do that my followup would be, are you sure we don't need to flush somewhere in the calling context ? But moving this into the undelete method I'm happy to just assume we do the right thing 😅

Reply:
Wait, why is it important to change the update time if no change was made ?

Reply:
It is because of the way we check for the completion of the bulk operation in the client... :disappointed: 
When we run the bulk operation the only way we have to tell it is finished (it might run in a background task) is by checking if the history update_time changed.

https://github.com/galaxyproject/galaxy/blob/4e2aaf8e2b88370e5805c02b1dd51bdff2bd0220/client/src/components/History/CurrentHistory/HistoryOperations/DefaultOperations.vue#L48

Maybe we could devise a better way of checking for the completion? I also don't like having to resort to this "hack".

Reply:
Can we hand back a handle to the task and poll the tasks API instead?

Reply:
Maybe we can explore that now, it was not an easy option when this was designed as Celery was more "optional/experimental" back then. It will require changes in the API though so I don't know if this will fit as a bug fix for 24.0 :thinking:

Reply:
So, what do you all think about leaving these "hacky" `item.update()` calls here in the bulk operations as a bug fix for 24.0 and trying to follow up in `dev` changing the API to handle asynchronous tasks? Then we can remove them from here.

Refreshing the `history.update_time` is still important when any of the items change after the bulk operation so the history watcher can fetch those changes. But there are cases when a bulk operation legitimately does not alter anything in the items because it doesn't apply (i.e. trying to undelete purged datasets). These cases should not trigger an update of the history, so to avoid the "forever running operation visual issue" we could track the status of the Celery task if the API returned the ID as Dannon suggested.

Reply:
I don't mind us fixing these like this for now.  The downside is pretty minimal, and, really, it only applies in a situation when the user really did want to make an update to the history anyway so I don't think anyone's going to care if the update time gets bumped.